### PR TITLE
SSD performance optimization and benchmark script

### PR DESCRIPTION
**Pull request by TaoLv**
## Description ##
In this PR:

- An inference benchmark script for SSD is added
- Operator MultiBoxDetection is optimized to improve SSD performance on CPU
- I tried to add a unit test for the consistency check between cpu and gpu but failed, since MultiBoxDetection on gpu cannot generate stable results for different run. I have raised a issue for this #10316. This code has past the consistency check with original cpu implementation.

`python benchmark_score.py`
SSD-VGG16 with 20 classes and shape (3, 300, 300), img/sec

bs | master | +mkldnn | +mkldnn+this PR
:--: | :--: | :--: | :--:
1 | 2.02 | 7.64 | 23.77
2 | 2.10 | 5.28 | 15.54
4 | 2.26 | 6.80 | 23.08
8 | 2.40 | 8.03 | 28.78
16 | 2.80 | 9.11 | 33.14
32 | 2.82 | 8.80 | 32.81

`python benchmark_score.py --network resnet50`
SSD-ResNet50 with 20 classes and shape (3, 300, 300), img/sec

bs | master | +mkldnn | +mkldnn+this PR
:--: | :--: | :--: | :--:
1 | 8.32 | 30.94 | 40.58
2 | 8.58 | 28.71 | 37.65
4 | 8.69 | 39.02 | 54.04
8 | 8.59 | 47.72 | 68.90
16 | 8.54 | 47.64 | 77.39
32 | 8.79 | 54.74 | 89.62

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by TaoLv** [comment]
@pengzhao-intel

**Comment by pengzhao-intel** [comment]
There's huge improvements for SSD.  
Is it possible to add this one into release 1.2?

**Comment by zhanghang1989** [comment]
@zhreshold

File: example/ssd/benchmark_score.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+from __future__ import print_function
+import os
+import sys
+import argparse
+import importlib
+import mxnet as mx
+import time
+#from dataset.iterator import DetRecordIter
+#from config.config import cfg
+#from evaluate.eval_metric import MApMetric, VOC07MApMetric
```
Comment (anirudh2290) [inline example/ssd/benchmark_score.py]: Remove commented code ?

File: src/operator/contrib/multibox_detection.cc [hunk 2]
```diff
@@ -112,20 +117,33 @@ inline void MultiBoxDetectionForward(const Tensor<cpu, 3, DType> &out,
...
-        // [id, prob, xmin, ymin, xmax, ymax]
-        p_out[valid_count * 6] = id - 1;  // remove background, restore original id
-        p_out[valid_count * 6 + 1] = (id == 0 ? DType(-1) : score);
-        int offset = i * 4;
-        TransformLocations(p_out + valid_count * 6 + 2, p_anchor + offset,
-          p_loc_pred + offset, clip, variances[0], variances[1],
-          variances[2], variances[3]);
+
+      // [id, prob, xmin, ymin, xmax, ymax]
+      outputs[i * 6] = id - 1;
```
Comment (anirudh2290) [inline src/operator/contrib/multibox_detection.cc]: Do any tests exist currently for this op ?

**Comment by zhreshold** [comment]
I hold the same opinion with @kevinthesun , it is not precise to benchmark randomly intialized weights.

**Comment by piiswrong** [comment]
So this would only help at the beginning of training?

**Comment by kevinthesun** [comment]
MultiBoxDetection is only used for inference. @zhreshold Is this accurate?

**Comment by zhreshold** [comment]
I think this should still improve the inference speed, but we would like to see how this affecting the pre-trained models. Apparently not by such a huge margin.

File: example/ssd/benchmark_score.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    num_batches = 100
+    dry_run = 5   # use 5 iterations to warm up
+    
+    for bs in batch_sizes:
+        batch = get_data(bs)
+        mod = mx.mod.Module(net, label_names=None, context=mx.cpu())
+        mod.bind(for_training = False,
+                 inputs_need_grad = False,
+                 data_shapes = get_data_shapes(bs))
+        mod.init_params(initializer=mx.init.Xavier(magnitude=2.))
```
Comment (zhreshold) [inline example/ssd/benchmark_score.py]: try load some pre-trained models to test the `real` perf

File: example/ssd/benchmark_score.py (hunk 1 above)
Comment (TaoLv) [reply example/ssd/benchmark_score.py]: Yes, will do.

File: src/operator/contrib/multibox_detection.cc (hunk 2 above)
Comment (TaoLv) [reply src/operator/contrib/multibox_detection.cc]: I guess no

File: example/ssd/benchmark_score.py (hunk 3 above)
Comment (TaoLv) [reply example/ssd/benchmark_score.py]: Sure. Do you know where I can find pre-trained parameters for SSD? I didn't find them in data.mxnet.io.

**Comment by pengzhao-intel** [comment]
@zhreshold @kevinthesun @piiswrong where we can get the pre-trained SSD data?

**Comment by juliusshufan** [comment]
@kevinthesun Yes, pre-trained model can be used for inference benchmarking, while it still involving dataloader and dataset, the benchmark script appended with this PR is dummy data-based and might be more flexible, I guess it is a potential utility as well. What do y ou think? :)

**Comment by kevinthesun** [comment]
@juliusshufan Synthetic data is fine as along as pre-trained model is used.

**Comment by zhreshold** [comment]
LGTM for this PR, just wish the benchmarks could be updated for the pretrained models so we can have better ideas of the perf improvement.
I will merge this now.
